app/compare/diff_preprocessing.tsx

```diff
@@ -0,0 +1,128 @@
+import { invocation as invocation_proto } from "../../proto/invocation_ts_proto";
+import { command_line } from "../../proto/command_line_ts_proto";
+
+/**
+ * A unique token used to identify parts of the diff that were hidden due to pre-processing options.
+ */
+export const HIDDEN_TOKEN = `/__HIDDEN__${Math.random()}/`;
+
+export type PreProcessingOptions = {
+  sortEvents?: boolean;
+  hideTimingData?: boolean;
+  hideConsoleOutput?: boolean;
+  hideInvocationIds?: boolean;
+  hideUuids?: boolean;
+};
+
+export function computeTextForDiff(
+  invocation: invocation_proto.IInvocation,
+  { sortEvents, hideTimingData, hideConsoleOutput, hideInvocationIds, hideUuids }: PreProcessingOptions = {}
+): string {
+  const invocationId = invocation.invocationId;
+  // Clone the invocation to avoid mutating the original object.
+  invocation = invocation_proto.Invocation.fromObject((invocation as invocation_proto.Invocation).toJSON());
+
+  if (sortEvents) {
+    sortByProperty(invocation.event, (event: any) => JSON.stringify(event?.buildEvent?.id));
+  }
+  if (hideConsoleOutput) {
+    invocation.consoleBuffer = HIDDEN_TOKEN;
+  }
+
+  // Some CommandLine objects are empty for some reason; remove these.
+  invocation.structuredCommandLine = invocation.structuredCommandLine.filter(
+    (commandLine: any) => commandLine.commandLineLabel
+  );
+  // Sort structured command lines so canonical always comes before original.
+  sortByProperty(invocation.structuredCommandLine, (commandLine) => commandLine.commandLineLabel);
+
+  if (hideTimingData) {
+    for (const commandLine of invocation.structuredCommandLine) {
+      removeTimingData(commandLine);
+    }
+  }
+
+  for (const event of invocation.event) {
+    const id = event?.buildEvent?.id;
+    if (!id) continue;
+
+    if (id?.configuration) {
+      // The "makeVariable" map sometimes shows the same data but rendered in a different order;
+      // sorting the maps solves this problem.
+      sortEntriesByKey(event.buildEvent.configuration.makeVariable);
+      continue;
+    }
+
+    if (hideTimingData) {
+      if (id?.workspaceStatus) {
+        event.buildEvent.workspaceStatus.item.find((item: any) => item.key === "BUILD_TIMESTAMP")!.value = HIDDEN_TOKEN;
+      } else if (id?.buildToolLogs) {
+        event.buildEvent.buildToolLogs.log.find(
+          (item: any) => item.name === "elapsed time"
+        )!.contents = new Uint8Array();
```

Comment:
why do we set this one to a Uint8Array rather than HIDDEN_TOKEN? also do we need the exclamation mark?

Reply:
Updated to `HIDDEN_TOKEN` - originally I saw that the type was `Uint8Array` and figured I was stuck with it. But I later learned that protobufjs supports assigning a string, and it will encode it to bytes (it's not quite base64; it's using some stuff from bytebuffer.js). See the new `HIDDEN_TOKEN_BYTES_ENCODING` stuff I added above